replica/compaction_group.hh

```diff
@@ -219,15 +219,54 @@ public:
     future<> split(compaction_group_list&, sstables::compaction_type_options::split opt);
 };
 
-using storage_group_vector = utils::chunked_vector<std::unique_ptr<storage_group>>;
+using storage_group_ptr = std::unique_ptr<storage_group>;
+using storage_group_vector = utils::chunked_vector<storage_group_ptr>;
+using storage_group_vector_ptr = lw_shared_ptr<storage_group_vector>;
 
 class storage_group_manager {
+protected:
+    // The compaction group list is only a helper for accessing the groups managed by the storage groups.
+    // The list entries are unlinked automatically when the storage group, they belong to, is removed.
+    mutable compaction_group_list _compaction_groups;
```

Comment:
Why mutable?

Reply:
I have a patch to get rid of this intrusive list. it was added for making integration easier, but it's dangerous as a compaction group alive but no longer in use could still be found in iteration. I don't remember why I had to make it mutable, but we'll remove it anyway.

Reply:
we can rely only on a foreach_compaction_group or parallel_foreach_compaction_group for iterations.

Reply:
@avikivity @bhalevy before I proceed with a patch (actually I have an ongoing one for it), I'd like to discuss this idea with you both. I am planning to rename storage_group to compaction_group_list, so we don't have to live with both terms (storage and compaction group). Think of compaction_group_list as our memtable_list. memtable_list may be composed of more than one memtable, all of which belong to the same replica table. Likewise, compaction_group_list is composed of more than one group, all of which belong to the same replica tablet. Makes sense, right? Then we can rename storage_group_manager to simply compaction_group_manager. Having both terms are very confusing, and making interfaces harder to reason about, as you don't know which one to work with.

Reply:
So replica::table will have a set of compaction group lists, where each list belongs to an individual tablet replica.

Reply:
storage_group_vector_ptr is replaced by compaction_group_lists = sparse_chunked_vector<compaction_group_list>; compaction_group_list = today's storage_group.

Reply:
Some more ideas:

1. splittable_compaction_group, compaction_group

2. compaction_group base class, simple_compaction_group implementation, splitting_compaction_group implementation (with base, left, right simple_compaction_group members)